src/quickgui/attributes/qgsquickattributemodel.cpp

```diff
@@ -0,0 +1,327 @@
+{
+  if ( mFeatureLayerPair.feature() == feature )
+    return;
+
+  beginResetModel();
+  mFeatureLayerPair = QgsQuickFeatureLayerPair( feature, mFeatureLayerPair.layer() );
+  endResetModel();
+
+  emit featureChanged();
+}
+
+QHash<int, QByteArray> QgsQuickAttributeModel::roleNames() const
+{
+  QHash<int, QByteArray> roles = QAbstractListModel::roleNames();
+  roles[AttributeName]  = QByteArrayLiteral( "AttributeName" );
+  roles[AttributeValue] = QByteArrayLiteral( "AttributeValue" );
+  roles[Field] = QByteArrayLiteral( "Field" );
+  roles[RememberAttribute] = QByteArrayLiteral( "RememberAttribute" );
+
+  return roles;
+}
+
+
+int QgsQuickAttributeModel::rowCount( const QModelIndex &parent ) const
+{
+  if ( parent.isValid() )
+    return 0;
+  else
+    return mFeatureLayerPair.feature().attributes().count();
+}
+
+QVariant QgsQuickAttributeModel::data( const QModelIndex &index, int role ) const
+{
+  switch ( role )
+  {
+    case AttributeName:
+      return mFeatureLayerPair.layer()->attributeDisplayName( index.row() );
+      break;
+
+    case AttributeValue:
+      return mFeatureLayerPair.feature().attribute( index.row() );
+      break;
+
+    case Field:
+      return mFeatureLayerPair.layer()->fields().at( index.row() );
+      break;
+
+    case RememberAttribute:
+      return mRememberedAttributes.at( index.row() );
+      break;
+  }
+
+  return QVariant();
+}
+
+bool QgsQuickAttributeModel::setData( const QModelIndex &index, const QVariant &value, int role )
+{
+  if ( data( index, role ) == value )
+    return true;
+
+  switch ( role )
+  {
+    case AttributeValue:
+    {
+      QVariant val( value );
+      QgsField fld = mFeatureLayerPair.feature().fields().at( index.row() );
+
+      if ( !fld.convertCompatible( val ) )
+      {
+        QgsMessageLog::logMessage( tr( "Value \"%1\" %4 could not be converted to a compatible value for field %2(%3)." ).arg( value.toString(), fld.name(), fld.typeName(), value.isNull() ? "NULL" : "NOT NULL" ) );
+        return false;
+      }
+      bool success = mFeatureLayerPair.featureRef().setAttribute( index.row(), val );
+      if ( success )
+        emit dataChanged( index, index, QVector<int>() << role );
+      return success;
+      break;
+    }
+
+    case RememberAttribute:
+    {
+      mRememberedAttributes[ index.row() ] = value.toBool();
+      emit dataChanged( index, index, QVector<int>() << role );
+      break;
+    }
+  }
+
+  return false;
+}
+
+bool QgsQuickAttributeModel::save()
+{
+  if ( !mFeatureLayerPair.layer() )
+    return false;
+
+  bool rv = true;
+
+  if ( !startEditing() )
+  {
+    rv = false;
+  }
+
+  QgsFeature feat = mFeatureLayerPair.feature();
+  if ( !mFeatureLayerPair.layer()->updateFeature( feat ) )
+    QgsMessageLog::logMessage( tr( "Cannot update feature" ),
+                               QStringLiteral( "QgsQuick" ),
+                               Qgis::Warning );
+  rv = commit();
```

Comment:
Might be good to find a better approach to dealing with this in the long run.

I assume, this calls lower-level I/O functions which shouldn't be used in a `Q_INVOKABLE` because they can make the UI unresponsive.